[WIP] AggLayer integration

nivida opened the pull request:
### What does this PR do?

It does add the option to settle the final proof over AggLayer with a soundness check. 

### Reviewers

Main reviewers:
- TBD

Codeowner reviewers:

- TBD

nivida committed:
draft integration of beethoven endpoint for proof settlement

nivida committed:
Getter added for RollupID to the etherman struct and updated the depending aggreagtor/aggregator.go code

nivida committed:
trusted sequencer private key added to cmd/run, aggregator/aggregator, and aggregator/config

nivida committed:
overtook aggregator_test changes from Arnau's PR

nivida committed:
etherman mock for aggregator test updated with mockery

nivida committed:
dependencies updated

nivida commented:
FYI: Beethoven is still a private repository and until we put it out to the public will those CI tasks fail.

Stefan-Ethernal commented on aggregator/aggregator.go:
```diff
@@ -267,34 +279,139 @@ func (a *Aggregator) sendFinalProof() {
-				mTxLogger.Errorf("Error to add batch verification tx to eth tx manager: %v", err)
-				a.handleFailureToAddVerifyBatchToBeMonitored(ctx, proof)
-				continue
+			switch a.cfg.SettlementBackend {
+			case Beethoven:
+				if success := a.settleWithBeethoven(ctx, proof, inputs); !success {
+					continue
+				}
+			default:
+				if success := a.settleDirect(ctx, proof, inputs); !success {
+					continue
+				}
 			}
```
WDYT about abstracting these different settlement strategies into some interface. Let's call it e.g. `ISettler` with the following function:

```go
type ISettler interface {
	Settle(ctx context.Context, proof *state.Proof, inputs types.FinalProofInputs) bool
}
```
You would have two implementations e.g. `L1Settler` and `BeethovenSettler` and instantiate the corresponding one based on the `SettlementBackend` "enumeration". The constructor function of the aggregator (`New`) would receive `ISettler` as a parameter and you could omit the trusted sequencer private key (`sequencerPrivateKey`) parameter, because it is not really needed by the aggregator per se, but only by Beethoven settling strategy. In that light, only `BeethovenSettler` would receive sequencer private key.

nivida replied:
Yep, totally agree. This is the correct way to implement that. On it.

nivida replied:
I've looked closer at the aggregator code and I think we can go with that switch/case cause this will otherwise end up in a full refactoring of it. Many basic architectural principles are violated here.

tclemos commented:
Since the PR is WIP, I'm converting it to draft until it is ready to be reviewed and merged.

vcastellm committed:
Merge remote-tracking branch 'origin' into feature/beethoven-integration

vcastellm committed:
Go get

agnusmor committed:
fix smtProof

vcastellm committed:
Go sum

vcastellm committed:
Merge tag 'v0.5.0-RC19' into feature/beethoven-integration

praetoriansentry committed:
fix: using constructor to create StackTrie

In some circumstances, we've encountered panics when using the empty
struct for the stacktrie because the hasher is nil. This commit
switches to using `NewSTackTrie` to create the object and avoids those
panics.

vcastellm committed:
Load PK from Keystore file config

Not sure if this was already implemented and can be reused.

nivida commented on config/config.go:
```diff
@@ -183,3 +187,19 @@ func Load(ctx *cli.Context, loadNetworkConfig bool) (*Config, error) {
 }
+
+// NewKeyFromKeystore creates a private key from a keystore file
+func NewKeyFromKeystore(cfg types.KeystoreFileConfig) (*ecdsa.PrivateKey, error) {
```
We could also already apply the related changes from Arnau here. [See here.](https://github.com/0xPolygon/cdk-validium-node/blob/a03bdeb526891222e22e138d2a83408b43559c25/etherman/etherman.go#L1739)

vcastellm committed:
Merge remote-tracking branch 'origin' into feature/beethoven-integration

vcastellm committed:
go mod tidy

vcastellm committed:
Update refs to agglayer

agnusmor committed:
update prover image to v4.0.2 (#3162)

joanestebanr committed:
Add a note in deployment instruction to restart rpc after forkId 7 (#3163)

* note to restart zkevm-rpc

* update comment

agnusmor committed:
change position of point 10

agnusmor committed:
add point 11

agnusmor committed:
add lines

agnusmor committed:
add breakline

agnusmor committed:
add lines

agnusmor committed:
add line

ARR552 committed:
Sync old network compatibility (#3166)

* Sync old network compatibility

* logs

joanestebanr committed:
synchronizer: unittest and remove pool_reorg  (#3111) (#3168)

* removed reorgPool call
* add test for l1 sequenced batches
* removed panic after halt call as PR comments ask for

agnusmor committed:
fix parentHash when storing a new L2 block (#3171)

ARR552 committed:
Fix/rom error (#3177)

* fix rom error

* linter

* fix

ARR552 committed:
Cherrypick/#3179 geth version v1.13.11 (#3180)

* new geth version v1.13.11

agnusmor committed:
Add sanity check: last L2 block timestamp in the sequence is also L1BlockTimestampMargin seconds behind "now" (#3181)

* add sanity check L1BlockTimestampMargin vs now

* Update prover image to v4.0.4

joanestebanr committed:
synchronizer: update fromTrusted fix cache update (#3169)

* fix cache update and sync from Trusted of closed batch

joanestebanr committed:
Feature/3173 sync trusted state in networks previous to  ETROG (#3175)

vcastellm committed:
Merge remote-tracking branch 'origin/develop' into feature/beethoven-integration

vcastellm committed:
Merge latest changes

joanestebanr committed:
Fix/3164 eth syncing dont return expected values (#3165)

ToniRamirezM committed:
fix im state root in stream (#3192)

joanestebanr committed:
sync: fix getting URL from contract in compatibility mode, and set sequential by default (#3191)

* fix getting URL from contract in compatibility mode
* synchronizer configuration  by default use `sequential` mode

agnusmor committed:
Fix isSynced function (#3188)

* fix isSynced

* fix linter

* fix logs

* halt on sanity check error

tclemos committed:
fix l2 block gas price to use value returned by the executor (#3186) (#3190)

tclemos committed:
allow rpc to access the ForkID directly from DB (#3196)

* allow rpc to access the ForkID directly from DB

* update docs

tclemos committed:
allow rpc to access the ForkID directly from DB (#3197)

* allow rpc to access the ForkID directly from DB

* update docs

* fix migration down; check for greater forkIDs instead of the next +1 (#3182)

* fix state storage config initialization

* add tests and fix get forkID by block number

* fix test TestGetForkIDByBlockNumber

---------

Co-authored-by: agnusmor <agnusmor@gmail.com>

joanestebanr committed:
Fix/3200 l1inforoot error on reorg (#3201)

* L1InfooTree cache is calculated when needed and is discard when a reorg happens

joanestebanr committed:
Fix/3203 panic if batch process fail and must be closed (#3206)

* synchronizer avoid panic if process trusted batch fails

agnusmor committed:
Remove L1SynchronizationMode parameter from environments config files (it will use default "sequential" value) (#3210)

tclemos committed:
allow GetL2TxHashByTxHash to return null for txs prior to Etrog (#3209)

agnusmor committed:
Fix L2 block gas limit to 2**50 (1125899906842624) (#3219)

* set l2 block gas limit to 2**50 (1125899906842624)

* limit L2 block gasLimit in StoreL2Block function

joanestebanr committed:
change 2^64-1 for an esmitation of blocks (#3220)

joanestebanr committed:
removed restart RPC on deploy instructions (#3221)

ToniRamirezM committed:
Cherry pick/3222 (#3224)

* fix stream

* fix stream

* fix stream

* fix stream

* linter

agnusmor committed:
add retries in is Synced function when last virtual batch is greater than last SC sequenced batcj (#3225)

joanestebanr committed:
Fix/3233 the etrog event UpdateEtrogSequence doesnt store the GER in batch table (#3234)

agnusmor committed:
remove state.batch section from environments config files (#3230)

ToniRamirezM committed:
get GER for l2block on Etrog (#3238)

agnusmor committed:
Cherry-pick #3240, #3137: Sequence batch only when sanity check is done. Do fatal for some finalizer halt events. Several fixes. Optmize purge pool (#3241)

* sequence batch only when sanity check is done. Do fatal for some halt events. Several fixes

* feat: optimize purge txn pool query (#3137)

* feat: optimize purge txn pool query

* fix: add new method to state interface

* feat: generate new state mocks

* fix merge

* fix %w

---------

Co-authored-by: Idris Hanafi <ifh101@gmail.com>

agnusmor committed:
fix node OOC (#3263)

vcastellm committed:
Merge tag 'v0.5.9' into feature/beethoven-integration

vcastellm committed:
oops
